crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -1099,6 +1099,46 @@ impl<'w> EntityWorldMut<'w> {
         unsafe { self.get_mut_assume_mutable() }
     }
 
+    /// Temporarily removes a [`Component`] `T` from this [`Entity`] and runs the
+    /// provided closure on it, returning the result if `T` was available.
+    /// This will trigger the `OnRemove` and `OnReplace` component hooks without
+    /// causing an archetype move.
+    ///
+    /// While this is available for all components, it's recommended to only be
+    /// used with immutable components.
+    /// When available, prefer using [`get_mut`](EntityMut::get_mut).
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use bevy_ecs::prelude::*;
+    /// #
+    /// #[derive(Component, PartialEq, Eq, Debug)]
+    /// #[component(immutable)]
+    /// struct Foo(bool);
+    ///
+    /// # let mut world = World::default();
+    /// # world.register_component::<Foo>();
+    /// #
+    /// # let entity = world.spawn(Foo(false)).id();
+    /// #
+    /// # let mut entity = world.entity_mut(entity);
+    /// #
+    /// # assert_eq!(entity.get::<Foo>(), Some(&Foo(false)));
+    /// #
+    /// entity.with_component(|foo: &mut Foo| {
+    ///     foo.0 = true;
+    /// });
+    /// #
+    /// # assert_eq!(entity.get::<Foo>(), Some(&Foo(true)));
+    /// ```
+    #[inline]
+    pub fn with_component<T: Component, R>(&mut self, f: impl FnOnce(&mut T) -> R) -> Option<R> {
+        self.world
+            .with_component(self.entity, f)
```

Comment:
I'd say your summary is pretty accurate! I added the method to `DeferredWorld` mostly because that was the most convenient one to trigger hooks/observers from! I think this is a fair concern, so I'll make the `DeferredWorld::with_component` `pub(crate)` as a precaution (can always make it public later). I've also made sure the `World` variant flushes before returning (if required).